autodiff/forward/real/real.hpp

```diff
@@ -656,6 +657,60 @@ constexpr auto atan(const Real<N, T>& x)
     return res;
 }
 
+template<size_t N, typename T, typename U, EnableIf<isArithmetic<U>>...>
+constexpr auto atan2(const U& c, const Real<N, T>& x)
+{    
+    // d[atan2(c,x)]/dx = -c / (c^2 + x^2)
+    Real<N, T> res;
+    res[0] = atan2(c, x[0]);
+    if constexpr(N > 0) {
+        Real<N - 1, T> xprime;
+        For<1, N + 1>([&](auto i) constexpr {
+            xprime[i - 1] = x[i];
+        });
+        Real<N - 1, T> aux(x);
+        aux = xprime * (-c / (c * c + aux * aux));
+        For<1, N + 1>([&](auto i) constexpr {
+            res[i] = aux[i - 1];
+        });
+    }
+    return res;
+}
+
+template<size_t N, typename T, typename U, EnableIf<isArithmetic<U>>...>
+constexpr auto atan2(const Real<N, T>& y, const U& c)
+{
+    // d[atan2(y,c)]/dy = c / (c^2 + y^2)
+    Real<N, T> res;
+    res[0] = atan2(y[0], c);
+    if constexpr(N > 0) {
+        Real<N - 1, T> yprime;
+        For<1, N + 1>([&](auto i) constexpr {
+            yprime[i - 1] = y[i];
+        });
+        Real<N - 1, T> aux(y);
+        aux = yprime * (c / (c * c + aux * aux));
+        For<1, N + 1>([&](auto i) constexpr {
+            res[i] = aux[i - 1];
+        });
+    }
+    return res;
+}
+
+template<size_t N, typename T>
+constexpr auto atan2(const Real<N, T>& y, const Real<N, T>& x)
+{
+    Real<N, T> res;
+    res[0] = atan2(y[0], x[0]);
+    if constexpr(N > 0) {
+        const T denom = x[0] * x[0] + y[0] * y[0];
+        For<1, N + 1>([&](auto i) constexpr {
+            res[i] = (x[0] * y[i] - x[i] * y[0]) / denom;
```

Comment:
@allanleal I am not entirely sure why this is true and was not able to derive it for myself. I mainly copied the pattern from  the implementation for `dual` seen [here](https://github.com/autodiff/autodiff/blob/main/autodiff/forward/dual/dual.hpp#L1403-L1407). If you could help me understand why this is valid I would greatly appreciate it.

Reply:
Thanks, @bostoncleek . I haven't yet published the algorithm behind `real` (just don't have time now). It's an algorithm different than the classic dual approach, specialized for single-order and higher-order directional derivatives.

Reply:
I'm very interested in the paper and will be awaiting its release. Are there similar papers you can refer me to or are you saying its a completely different algorithm?